File: OrderMatchingEngine/src/main/java/com/ndgndg91/ordermatchingengine/order/MatchedEntry.java
package com.ndgndg91.ordermatchingengine.order;

import lombok.Getter;
import lombok.ToString;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Getter
@ToString
public class MatchedEntry {
    private final String orderId;
    private final OrderType orderType;
    private final int shares;
    private final PriceType priceType;
    private final BigDecimal price;
    private final LocalDateTime timestamp;

    public MatchedEntry(OrderEntry e, int shares) {
        this.orderId = e.getOrderId();
        this.orderType = e.getOrderType();
        this.shares = shares;
        this.priceType = e.getPriceType();
        this.price = e.getPrice();
        this.timestamp = e.getTimestamp();
    }

    public MatchedEntry(OrderEntry e, String orderId) {
        this.orderId = e.getOrderId();
        this.orderType = e.getOrderType();
        this.shares = e.partialShares(orderId);
        this.priceType = e.getPriceType();
        this.price = e.getPrice();
        this.timestamp = e.getTimestamp();
    }

    public BigDecimal totalPrice() {
        return price.multiply(BigDecimal.valueOf(shares));
    }
}

File: OrderMatchingEngineKotlin/src/main/kotlin/com/ndgndg91/ordermatchingenginekotlin/order/MatchedEntry.kt
package com.ndgndg91.ordermatchingenginekotlin.order

import java.math.BigDecimal
import java.time.LocalDateTime

class MatchedEntry {
    val orderId: String
    val orderType: OrderType
    val shares: Int
    val priceType: PriceType
    val price: BigDecimal
    val timestamp: LocalDateTime

    constructor(entry: OrderEntry, s: Int) {
        this.orderId = entry.orderId
        this.orderType = entry.orderType
        this.shares = s
        this.priceType = entry.priceType
        this.price = entry.price
        this.timestamp = entry.timestamp
    }

    constructor(entry: OrderEntry, orderId: String) {
        this.orderId = entry.orderId
        this.orderType = entry.orderType
        this.shares = entry.partialShares(orderId)
        this.priceType = entry.priceType
        this.price = entry.price
        this.timestamp = entry.timestamp
    }

    fun totalPrice(): BigDecimal {
        return this.price.multiply(BigDecimal(this.shares))
    }
}

